File: managers/solutions/ManagersBranch.java
import java.io.*;

import java.util.List;
import java.util.ArrayList;

public class ManagersBranch {
    static long solve(int N, int[] M) {
        List<List<Integer>> subordinates = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            subordinates.add(new ArrayList<>());
        }
        for (int i = 1; i < N; i++) {
            subordinates.get(M[i]).add(i);
        }
        
        int branchSizes[] = new int[N];
        findBranchSizeDFS(0, subordinates, branchSizes);

        // Since there can be up to 10 ** 5 employees, there can be up to ~10 ** 10
        // unique disputes. To represent the worst case where all of them are resolved
        // by the CEO, we need to use 64 bits instead.
        long mostDisputes = 0;
        for (int i = 0; i < N; i++) {
            mostDisputes = Math.max(mostDisputes, disputesResolvedBy(i, subordinates, branchSizes));
        }
        return mostDisputes;
    }

    static int findBranchSizeDFS(int id, List<List<Integer>> subordinates, int[] branchSizes) {
        branchSizes[id] = 1;
        for (int sub : subordinates.get(id)) {
            branchSizes[id] += findBranchSizeDFS(sub, subordinates, branchSizes);
        }
        return branchSizes[id];
    }

    static long disputesResolvedBy(int id, List<List<Integer>> subordinates, int[] branchSizes) {
        if (subordinates.get(id).size() == 0) {
            return 0;
        } else if (subordinates.get(id).size() == 1) {
            long selfDisputes = branchSizes[id] - 1;
            return selfDisputes;
        } else {
            long selfDisputes = branchSizes[id] - 1;

            long leftBranchSize = branchSizes[subordinates.get(id).get(0)];
            long rightBranchSize = branchSizes[subordinates.get(id).get(1)];
            long crossBranchDisputes = leftBranchSize * rightBranchSize;

            return selfDisputes + crossBranchDisputes;
        }
    }
    
    static BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
    static PrintWriter out = new PrintWriter(System.out);

    public static void main(String[] args) throws IOException {
        int T = Integer.parseInt(in.readLine());
        for (int i = 0; i < T; i++) {
            int N = Integer.parseInt(in.readLine());
            String info[] = in.readLine().split(" ");
            int[] M = new int[info.length];
            for (int j = 0; j < info.length; j++) {
                M[j] = Integer.parseInt(info[j]);
            }
            System.out.println(solve(N, M));
        }
        out.flush();
    }
}

File: managers/solutions/managers_branch.py
from typing import List

def solve(N: int, M: List[int]) -> int:
    subordinates = [[] for _ in range(N)]
    for i in range(1, N):
        subordinates[M[i]].append(i)
    
    branch_sizes = [None for _ in range(N)]
    def find_branch_size_dfs(id):
        branch_sizes[id] = 1
        for sub in subordinates[id]:
            branch_sizes[id] += find_branch_size_dfs(sub)
        return branch_sizes[id]
    find_branch_size_dfs(0)
    
    def disputes_resolved_by(id):
        if len(subordinates[id]) == 0:
            return 0
        elif len(subordinates[id]) == 1:
            self_disputes = branch_sizes[id] - 1
            return self_disputes
        else:
            self_disputes = branch_sizes[id] - 1
            
            left_branch_size = branch_sizes[subordinates[id][0]]
            right_branch_size = branch_sizes[subordinates[id][1]]
            cross_branch_disputes = left_branch_size * right_branch_size
            
            return self_disputes + cross_branch_disputes
    
    return max(disputes_resolved_by(i) for i in range(N))

def main():
    T = int(input())
    for _ in range(T):
        N = int(input().strip())
        M = [int(x) for x in input().strip().split(' ')]
        print(solve(N, M))

if __name__ == '__main__':
    main()
